File: server/internal/models/user_action.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserAction struct {
	ID     uuid.UUID
	Name   Action
	UserID uuid.UUID
}

type UserActionVerification struct {
	Action    UserAction
	Token     string
	ExpiresAt time.Time
	UsedAt    *time.Time
	Verified  bool
}

func (ua *UserAction) Response() *responses.UserActionResponse {
	return &responses.UserActionResponse{
		ID:         ua.ID.String(),
		ActionName: ua.Name.String(),
	}
}

func (ua *UserAction) DBCreate() *storage.CreateUserActionParams {
	return &storage.CreateUserActionParams{
		ID:     ua.ID,
		Name:   ua.Name.String(),
		UserID: ua.UserID,
	}
}

func (uv *UserActionVerification) DBCreate() *storage.CreateUserVerificationParams {
	return &storage.CreateUserVerificationParams{
		ActionID:  uv.Action.ID,
		Token:     uv.Token,
		ExpiresAt: uv.ExpiresAt.UnixMilli()}
}

func UserActionFromDB(row *storage.DivaAction) *UserAction {
	return &UserAction{
		ID:     row.ID,
		Name:   ActionFromString(row.Name),
		UserID: row.UserID,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/user/actions/UserAction.kt
package io.github.juevigrace.diva.lib.models.user.actions

import io.github.juevigrace.diva.core.Option
import io.github.juevigrace.diva.lib.models.actions.Actions
import io.github.juevigrace.diva.lib.models.actions.safeActionsValueOf
import io.github.juevigrace.diva.lib.models.api.user.action.UserActionResponse
import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class, ExperimentalJsExport::class)
@JsExport
data class UserAction(
    val id: Uuid,
    val action: Actions,
    val userId: Uuid = Uuid.NIL,
) {
    companion object {
        fun fromResponse(response: UserActionResponse): UserAction {
            return UserAction(
                id = Uuid.parse(response.id),
                action = safeActionsValueOf(response.actionName),
            )
        }
    }
}

@OptIn(ExperimentalJsExport::class)
@JsExport
data class UserActionVerification(
    val action: UserAction,
    val token: String,
    val expiresAt: Instant,
    val usedAt: Option<Instant> = Option.None,
    val verified: Boolean = false,
)
